internal/api/kubernetes/client.go

```diff
@@ -510,93 +522,174 @@ func gvrAndKeyFromObj(
-	ctx context.Context,
-	internalClient libClient.Client,
-	verb string,
-	gvr schema.GroupVersionResource,
-	subresource string,
-	key libClient.ObjectKey,
+func getAuthorizedClient(kargoNamespace string) func(
+	context.Context,
+	libClient.Client,
+	string,
+	schema.GroupVersionResource,
+	string,
+	libClient.ObjectKey,
 ) (libClient.Client, error) {
-	userInfo, ok := user.InfoFromContext(ctx)
-	if !ok {
-		return nil, errors.New("not allowed")
-	}
-
-	// Admins get to use the Kargo API server's own Kubernetes client. i.e. They
-	// can do everything the server can do.
-	if userInfo.IsAdmin {
-		return internalClient, nil
-	}
-
-	if userInfo.Username != "" {
-		// TODO: We are not yet mapping users to Kubernetes SAs, so anyone we could
-		// authenticate currently gets all the same permissions as the Kargo API
-		// server itself.
-		//
-		// We should replace this with a SubjectAccessReview on the SA to determine
-		// if the desired operation is allowed. If it is, we can return the
-		// internalClient.
-		return internalClient, nil
-	}
+	return func(
+		ctx context.Context,
+		internalClient libClient.Client,
+		verb string,
+		gvr schema.GroupVersionResource,
+		subresource string,
+		key libClient.ObjectKey,
+	) (libClient.Client, error) {
+		userInfo, ok := user.InfoFromContext(ctx)
+		if !ok {
+			return nil, errors.New("not allowed")
+		}
 
-	// If we get to here, we're dealing with a user who "authenticated" by just
-	// passing their bearer token for the Kubernetes API server.
+		// Admins get to use the Kargo API server's own Kubernetes client. i.e. They
+		// can do everything the server can do.
+		if userInfo.IsAdmin {
+			return internalClient, nil
+		}
 
-	// Calling getUserClient will build a user-specific client and conduct a
-	// SelfSubjectAccessReview. If the review succeeds, we'll have obtained a
-	// client that can perform the desired operation. If the review fails, the
-	// returned client will be nil.
-	client, err := getUserClient(
-		ctx,
-		internalClient.Scheme(),
-		userInfo.BearerToken,
-		authv1.ResourceAttributes{
+		ra := authv1.ResourceAttributes{
 			Verb:        verb,
 			Group:       gvr.Group,
 			Version:     gvr.Version,
 			Resource:    gvr.Resource,
 			Subresource: subresource,
 			Namespace:   key.Namespace,
 			Name:        key.Name,
-		},
-	)
-	if err != nil {
-		return nil,
-			errors.Wrap(err, "error getting user-specific Kubernetes client")
+		}
+		if userInfo.Username != "" {
+			serviceAccounts := make([]types.NamespacedName, 0)
+			for _, ns := range []string{
+				key.Namespace,
+				kargoNamespace,
+			} {
+				if ns == "" {
+					continue
+				}
+				accounts, ok := userInfo.ServiceAccounts[ns]
+				if !ok {
+					continue
+				}
+				for _, name := range accounts {
+					serviceAccounts = append(serviceAccounts, types.NamespacedName{
+						Namespace: ns,
+						Name:      name,
+					})
+				}
+			}
+			// If the operation is cluster-scoped, all ServiceAccounts are candidates.
```

Comment:
This I am not sure about...

I don't think the API server deals with _any_ cluster-scoped resources, so I'd rather just leave this out for now and consider putting it back later if we figure out it's needed.

Reply:
I forgot about namespace. You're right. Leave this, but let's continue putting more thought into it.